RFS/src/main/java/com/rfs/common/http/RestClient.java

```diff
@@ -0,0 +1,185 @@
+        public final String message;
+        public final Map<String, String> headers;
+
+        public Response(int responseCode, String responseBody, String responseMessage, Map<String, String> responseHeaders) {
+            this.code = responseCode;
+            this.body = responseBody;
+            this.message = responseMessage;
+            this.headers = responseHeaders;
+        }
+    }
+
+    public final ConnectionContext connectionContext;
+    private final HttpClient client;
+    private final AuthTransformer authTransformer;
+
+    @SneakyThrows
+    public RestClient(ConnectionContext connectionContext) {
+        this.connectionContext = connectionContext;
+        this.authTransformer = connectionContext.getAuthTransformer();
+
+        SslProvider sslProvider;
+        if (connectionContext.insecure) {
+            SslContext sslContext = SslContextBuilder.forClient()
+                .trustManager(InsecureTrustManagerFactory.INSTANCE)
+                .build();
+
+            sslProvider = SslProvider.builder().sslContext(sslContext).handlerConfigurator(sslHandler -> {
+                SSLEngine engine = sslHandler.engine();
+                SSLParameters sslParameters = engine.getSSLParameters();
+                sslParameters.setEndpointIdentificationAlgorithm(null);
+                engine.setSSLParameters(sslParameters);
+            }).build();
+        } else {
+            sslProvider = SslProvider.defaultClientProvider();
+        }
+
+        this.client = HttpClient.create().secure(sslProvider).baseUrl(connectionContext.url);
+    }
+
+    public static String getHostHeader(String urlString) {
+        try {
+            URL url = new URL(urlString);
+            String host = url.getHost();
+            int port = url.getPort();
+            String protocol = url.getProtocol();
+
+            if (port == -1 || (protocol.equals("http") && port == 80) || (protocol.equals("https") && port == 443)) {
+                return host;
+            } else {
+                return host + ":" + port;
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
+            return null;
+        }
+    }
+
+    public Mono<Response> asyncRequestWithStringHeaderValues(HttpMethod method, String path, String body, Map<String, String> additionalHeaders,
+                                       IRfsContexts.IRequestContext context) {
+        var convertedHeaders = additionalHeaders.entrySet().stream().collect(Collectors
+            .toMap(Map.Entry::getKey, e -> List.of(e.getValue())));
+        return asyncRequest(method, path, body, convertedHeaders, context);
+    }
+
+
+    public Mono<Response> asyncRequest(HttpMethod method, String path, String body, Map<String, List<String>> additionalHeaders,
+                                       @Nullable IRfsContexts.IRequestContext context) {
+        Map<String, List<String>> headers = new HashMap<>(Map.of(
+            "User-Agent", List.of("RfsWorker-1.0"),
```

Comment:
This agent string seems like it should come from the irequest context

Reply:
Request context is not always present. Discussed offline and will keep it as it is for now